scripts/externalTests/common.sh

```diff
@@ -572,15 +572,12 @@ function gas_report_to_json
 
 function detect_hardhat_artifact_dir
 {
-    if [[ -e build/ && -e artifacts/ ]]; then
-        fail "Cannot determine Hardhat artifact location. Both build/ and artifacts/ exist"
-    elif [[ -e build/ ]]; then
-        echo -n build/artifacts
-    elif [[ -e artifacts/ ]]; then
-        echo -n artifacts
-    else
-        fail "Hardhat build artifacts not found."
-    fi
+    # NOTE: The artifacts path is a configured parameter in Hardhat, so the below may fail for new external tests
+    # See: https://hardhat.org/hardhat-runner/docs/config#path-configuration
+    local artifact_dir=()
+    [[ -e build/artifacts ]] && artifact_dir+=("build/artifacts")
+    [[ -e artifacts/ ]] && artifact_dir+=("artifacts")
+    echo -n "${artifact_dir[@]}"
```

Comment:
I'm not sure this this fixes things.

I see that the issue with the original code is that `fail` does not actually interrupt the script. Not sure why exactly. Maybe we should make `fail` do `exit` instead of `return`... The reason it does not do that now is that `exit` exits unconditionally, so it won't let you chain things that fail this way using `||`, which is sometimes handy and I think we rely on that in a few places. Maybe we should just define a variant like `hardFail` that does exit for those cases where the failure represents something closer to an assert than a recoverable error :)

But anyway, the current version differs only in that it never fails but instead may return one or both of these dirs. How does it fix ENS? Doesn't it just mean that now we now fail to detect its artifact dir and just do not inspect its artifacts?

Reply:
> I'm not sure this this fixes things.
> 
> I see that the issue with the original code is that `fail` does not actually interrupt the script. Not sure why exactly. Maybe we should make `fail` do `exit` instead of `return`... The reason it does not do that now is that `exit` exits unconditionally, so it won't let you chain things that fail this way using `||`, which is sometimes handy and I think we rely on that in a few places. Maybe we should just define a variant like `hardFail` that does exit for those cases where the failure represents something closer to an assert than a recoverable error :)
>
> But anyway, the current version differs only in that it never fails but instead may return one or both of these dirs. How does it fix ENS? Doesn't it just mean that now we now fail to detect its artifact dir and just do not inspect its artifacts?

Because ENS has both artifact locations, namely `./build/artifacts` and `./artifacts/`, my change now utilizes both if they exist, and it does nothing if they don't. Therefore, we don't really need to hard fail in the latter case, since no artifacts were found, and we can just let the tests fail by itself.

However, it's possible that artifacts may be located elsewhere, and to locate them, we would need to search the project path for any artifacts directory and assume that they are generated by Hardhat, which may not always be accurate and could lead to other failures.

I believe it's better to take no action if we cannot find artifacts where we expect them and allow the test to fail if they exist in a different path and for some reason cause issues.

Reply:
Ah, ok. I assumed that in case of ENS if was hitting the case with both dirs missing. This makes more sense.

In that case it sounds like returning two dirs indeed is the right fix.

> I believe it's better to take no action if we cannot find artifacts where we expect them and allow the test to fail if they exist in a different path and for some reason cause issues.

But will it really fail in that case? The reason we have this check in the first place is because the test won't automatically fail with wrong artifacts. It will happily continue even if they were built by a different version of solc. We force it to fail so that we can notice when the test is not really running with bytecode built by the custom-built solc binary we provide.

If neither of these two dirs exists, we should still hard fail, because the artifacts are probably still somewhre in there, just in a different location and we need to add that new location to the function.

Reply:
> But will it really fail in that case? The reason we have this check in the first place is because the test won't automatically fail with wrong artifacts. It will happily continue even if they were built by a different version of solc. We force it to fail so that we can notice when the test is not really running with bytecode built by the custom-built solc binary we provide.
> 
> If neither of these two dirs exists, we should still hard fail, because the artifacts are probably still somewhre in there, just in a different location and we need to add that new location to the function.

Oh, I see, I thought that it would fail if it used a wrong artifact. So yeah, in neither location is found we should thrown an error indeed. I will update the script.

Reply:
Great. After you add this back the PR will be fine to merge.

Reply:
Done, I just added this line: `(( ${#artifact_dir[@]} != 0 )) || assertFail`